### PR TITLE
Support DataRepresentationQos [19833]

### PR DESCRIPTION
**Pull request by richiware**
## Description

Support Datawriter and Datareader DataRepresentationQos policy.

Depends on:

- eprosima/fast-dds#3971

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- *N/A* Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- *N/A* New feature has been documented/Current behavior is correctly described in the documentation. 
    
- *N/A* Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: test/blackbox/common/BlackboxTestsDataRepresentationQos.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+#include <vector>
+#include <tuple>
+
+#include <gtest/gtest.h>
+
+#include <fastcdr/Cdr.h>
+
+#include "PubSubReader.hpp"
+#include "PubSubWriter.hpp"
```
Comment (adriancampo) [inline test/blackbox/common/BlackboxTestsDataRepresentationQos.cpp:25]: I think this is using the  wrong headers from `fastrtps_deprecated`.

Should be:
#include "../api/dds-pim/PubSubReader.hpp"
#include "../api/dds-pim/PubSubWriter.hpp"

Might be worth checking this for other tests.

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
@richiprosima please test mac

**Comment by richiware** [comment]
@richiprosima please test this

File: test/blackbox/common/BlackboxTestsDataRepresentationQos.cpp (hunk 1 above)
Comment (MiguelCompany) [reply test/blackbox/common/BlackboxTestsDataRepresentationQos.cpp]: We have two different projects, one for each API.
The CMake files take care of adding the corresponding api folder to the list of include flags.

File: src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp [hunk 2]
```diff
@@ -980,6 +984,12 @@ bool EDP::valid_matching(
         EPROSIMA_LOG_WARNING(RTPS_EDP, "Incompatible liveliness kinds: offered kind is < than requested kind");
         incompatible_qos.set(fastdds::dds::LIVELINESS_QOS_POLICY_ID);
     }
+    // DataRepresentationQosPolicy
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp:987]: It looks like this whole method could be replaced by 
```
    return valid_matching(wdata, rdata, reason, incompatible_qos);
```

File: test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (MiguelCompany) [inline test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp:1]:
```suggestion
// Copyright 2023 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

File: test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        if (std::get<0>(GetParam()).empty() || XCDR_DATA_REPRESENTATION == std::get<0>(GetParam()).at(0))
+        {
+            ASSERT_EQ(XCDR_DATA_REPRESENTATION, writer_type->last_data_representation);
+            ASSERT_EQ(eprosima::fastcdr::EncodingAlgorithmFlag::PLAIN_CDR, reader_type->last_encoding);
+        }
+        else if (XCDR2_DATA_REPRESENTATION == std::get<0>(GetParam()).at(0))
+        {
+            ASSERT_EQ(XCDR2_DATA_REPRESENTATION, writer_type->last_data_representation);
+            ASSERT_EQ(eprosima::fastcdr::EncodingAlgorithmFlag::PLAIN_CDR2, reader_type->last_encoding);
+        }
```
Comment (MiguelCompany) [inline test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp:128]: We could add the expected values to the test parameters, thus simplifying these lines.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Apart from my comments below, this does not build with fastcdr 1.1

File: test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp (hunk 4 above)
Comment (richiware) [reply test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp]: Done in bbf761728

File: test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp (hunk 3 above)
Comment (richiware) [reply test/blackbox/common/DDSBlackboxTestsDataRepresentationQos.cpp]: Done in bbf761728

File: src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp [hunk 5]
```diff
@@ -843,29 +850,26 @@ bool EDP::checkDataRepresentationQos(
...
+        else if (writerRepresentation == fastdds::dds::XCDR_DATA_REPRESENTATION)
+        {
+            compatible |= std::find(rr.begin(), rr.end(), fastdds::dds::XCDR_DATA_REPRESENTATION) != rr.end()
+                    || rr.empty();
+        }
+        else // XML_DATA_REPRESENTATION
+        {
+            EPROSIMA_LOG_INFO(EDP, "DataRepresentationQosPolicy XML_DATA_REPRESENTATION isn't supported.");
         }
+
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp]: Done in bbf761728

File: src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp [hunk 6]
```diff
@@ -843,29 +850,26 @@ bool EDP::checkDataRepresentationQos(
 
     if (wdata->m_qos.representation.m_value.empty())
     {
-        compatible |= std::find(rr.begin(), rr.end(), fastdds::dds::XCDR2_DATA_REPRESENTATION) != rr.end();
         compatible |= std::find(rr.begin(), rr.end(), fastdds::dds::XCDR_DATA_REPRESENTATION) != rr.end() || rr.empty();
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp]: Done in bbf761728

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp [hunk 7]
```diff
@@ -958,7 +958,10 @@ ReturnCode_t DataWriterImpl::perform_create_new_change(
             return ReturnCode_t::RETCODE_OUT_OF_RESOURCES;
         }
 
-        if ((ALIVE == change_kind) && !type_->serialize(data, &payload.payload))
+        if ((ALIVE == change_kind) && !type_->serialize(data, &payload.payload,
+                qos_.representation().m_value.empty()
+                || XCDR_DATA_REPRESENTATION == qos_.representation().m_value.at(0)
+                    ? XCDR_DATA_REPRESENTATION : XCDR2_DATA_REPRESENTATION))
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/publisher/DataWriterImpl.cpp]: Done in bbf761728

File: src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp (hunk 2 above)
Comment (richiware) [reply src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp]: Done in 9a5c3431d

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
Fix compilation against fastcdr v1 in  57864f3b10a62d855226fc69b441deb85e37cddb

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp [hunk 8]
```diff
@@ -181,6 +181,11 @@ DataWriterImpl::DataWriterImpl(
         is_custom_payload_pool_ = true;
         payload_pool_ = payload_pool;
     }
+
+    // Set Datawriter's DataRepresentationId taking into account the QoS.
+    data_representation_ = qos_.representation().m_value.empty()
```
Comment (MiguelCompany) [inline src/cpp/fastdds/publisher/DataWriterImpl.cpp:186]: Move this into the `enable()` method

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp [hunk 9]
```diff
@@ -958,7 +963,8 @@ ReturnCode_t DataWriterImpl::perform_create_new_change(
             return ReturnCode_t::RETCODE_OUT_OF_RESOURCES;
         }
 
-        if ((ALIVE == change_kind) && !type_->serialize(data, &payload.payload))
+        if ((ALIVE == change_kind) && !type_->serialize(data, &payload.payload,
+                data_representation_))
```
Comment (MiguelCompany) [inline src/cpp/fastdds/publisher/DataWriterImpl.cpp:967]:
```suggestion
        if ((ALIVE == change_kind) && !type_->serialize(data, &payload.payload, data_representation_))
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp (hunk 8 above)
Comment (richiware) [reply src/cpp/fastdds/publisher/DataWriterImpl.cpp]: Done in cbd45b9

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp (hunk 9 above)
Comment (richiware) [reply src/cpp/fastdds/publisher/DataWriterImpl.cpp]: Done in de87081

**Review by MiguelCompany — DISMISSED** [review]

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Review by MiguelCompany — APPROVED** [review]

**Comment by MiguelCompany** [comment]
@richiware Please rebase and fix the conflict on `versions.md` here

**Comment by richiware** [comment]
> @richiware Please rebase and fix the conflict on `versions.md` here

@MiguelCompany Done

**Pull request merged**